File: +tom/MeasureDefault.m
classdef MeasureDefault < tom.Measure
% Copyright 2011 Scientific Systems Company Inc., New BSD License
 
  methods (Static = true, Access = public)
    function initialize(name)
      function text = componentDescription
        text = 'This default measure constructs no graph edges.';
      end
      tom.Measure.connect(name, @componentDescription, @tom.MeasureDefault);
    end
  end
  
  methods (Access = public, Static = true)
    function this = MeasureDefault(initialTime, uri)
      this = this@tom.Measure(initialTime, uri);
    end
  end
  
  methods (Access = public)
    function refresh(this, x)
      assert(isa(this, 'hidi.Sensor'));
      assert(isa(x, 'tom.Trajectory'));
    end
    
    function flag = hasData(this)
      assert(isa(this, 'hidi.Sensor'));
      flag = false;
    end
    
    function n = first(this)
      assert(isa(this, 'hidi.Sensor'));
      n = uint32(0);
      assert(isa(n, 'uint32'));
      error('The default sensor has no data.');
    end
    
    function n = last(this)
      assert(isa(this, 'hidi.Sensor'));
      n = uint32(0);
      assert(isa(n, 'uint32'));
      error('The default sensor has no data.');
    end
    
    function time = getTime(this, n)
      assert(isa(this, 'hidi.Sensor'));
      assert(isa(n, 'uint32'));
      time = 0.0;
      assert(isa(time, 'double'));
      error('The default sensor has no data.');
    end
    
    function edgeList = findEdges(this, naMin, naMax, nbMin, nbMax)
      assert(isa(this, 'tom.Measure'));
      assert(isa(naMin, 'uint32'));
      assert(isa(naMax, 'uint32'));
      assert(isa(nbMin, 'uint32'));
      assert(isa(nbMax, 'uint32'));
      edgeList = repmat(tom.GraphEdge, [0, 1]);
    end
    
    function cost = computeEdgeCost(this, x, graphEdge)
      assert(isa(this, 'tom.Measure'));
      assert(isa(x, 'tom.Trajectory'));
      assert(isa(graphEdge, 'tom.GraphEdge'));
      cost = 0;      
    end
  end
  
end

File: +tom/MeasureDefault.cpp
// Copyright 2011 Scientific Systems Company Inc., New BSD License
#include "Measure.h"

namespace tom
{
  /** This is the default measure. It has no data and constructs no graph edges. */
  class MeasureDefault : public virtual Measure
  {
  public:
    /** Constructor and parent class initializer of the same form. */
    MeasureDefault(const double initialTime, const std::string uri) :
      Measure(initialTime, uri)
    {
      return;
    }

    void refresh(Trajectory* x)
    {
      return;
    }

    bool hasData(void)
    {
      return (false);
    }

    uint32_t first(void)
    {
      throw("The default sensor has no data.");
      return (0);
    }

    uint32_t last(void)
    {
      throw("The default sensor has no data.");
      return (0);
    }

    double getTime(const uint32_t& n)
    {
      throw("The default sensor has no data.");
      return (0);
    }

    void findEdges(const uint32_t naMin, const uint32_t naMax, const uint32_t nbMin, const uint32_t nbMax, std::vector<
        GraphEdge>& edgeList)
    {
      edgeList.resize(0);
      return;
    }

    double computeEdgeCost(Trajectory* x, const GraphEdge graphEdge)
    {
      return (0.0);
    }

  private:
    static std::string componentDescription(void)
    {
      return ("This is the default measure. It has no data and constructs no graph edges.");
    }

    static Measure* componentFactory(const double initialTime, const std::string uri)
    {
      return (new MeasureDefault(initialTime, uri));
    }

  protected:
    static void initialize(std::string name)
    {
      connect(name, componentDescription, componentFactory);
    }
    friend class MeasureDefaultInitializer;
  };

  /** This class initializes MeasureDefault before the execution of main(). */
  class MeasureDefaultInitializer
  {
  public:
    MeasureDefaultInitializer(void)
    {
      MeasureDefault::initialize("tom");
    }
  } _MeasureDefaultInitializer;
}

